cpp/src/arrow/compute/row/row_encoder_internal.h

```diff
@@ -269,6 +272,190 @@ struct ARROW_EXPORT NullKeyEncoder : KeyEncoder {
   }
 };
 
+template <typename ListType>
+struct ARROW_EXPORT ListKeyEncoder : KeyEncoder {
+  static_assert(is_list_like_type<ListType>(), "ListKeyEncoder only supports ListType");
+  using Offset = typename ListType::offset_type;
+
+  ListKeyEncoder(std::shared_ptr<DataType> self_type,
+                 std::shared_ptr<DataType> element_type,
+                 std::shared_ptr<KeyEncoder> element_encoder)
+      : self_type_(std::move(self_type)),
+        element_type_(std::move(element_type)),
+        element_encoder_(std::move(element_encoder)) {}
+
+  void AddLength(const ExecValue& data, int64_t batch_length, int32_t* lengths) override {
+    if (data.is_array()) {
+      ARROW_DCHECK_EQ(data.array.length, batch_length);
+      const uint8_t* validity = data.array.buffers[0].data;
+      const auto* offsets = data.array.GetValues<Offset>(1);
+      // AddLength for each list
+      std::vector<int32_t> child_lengthes;
```

Comment:
This is for `AddLength`, if lots of value being used, `AddLength` call `AddLength(child_lengthes.data(), length)` rather than call with length 1